google/resource_storage_notification.go

```diff
@@ -0,0 +1,162 @@
+package google
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+	"google.golang.org/api/storage/v1"
+)
+
+func resourceStorageNotification() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceStorageNotificationCreate,
+		Read:   resourceStorageNotificationRead,
+		Delete: resourceStorageNotificationDelete,
+		Importer: &schema.ResourceImporter{
+			State: resourceStorageNotificationImportState,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"bucket": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"payload_format": &schema.Schema{
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringInSlice([]string{"JSON_API_V1", "NONE"}, false),
+			},
+
+			"topic": &schema.Schema{
+				Type:             schema.TypeString,
+				Required:         true,
+				ForceNew:         true,
+				DiffSuppressFunc: compareSelfLinkOrResourceName,
+			},
+
+			"custom_attributes": &schema.Schema{
+				Type:     schema.TypeMap,
+				Optional: true,
+				ForceNew: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+
+			"event_types": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
+				ForceNew: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+					ValidateFunc: validation.StringInSlice([]string{
+						"OBJECT_FINALIZE", "OBJECT_METADATA_UPDATE", "OBJECT_DELETE", "OBJECT_ARCHIVE"},
+						false),
+				},
+			},
+
+			"object_name_prefix": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"self_link": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceStorageNotificationCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	bucket := d.Get("bucket").(string)
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	computedTopicName := getComputedTopicName(project, d.Get("topic").(string))
+
+	storageNotification := &storage.Notification{
+		CustomAttributes: expandStringMap(d, "custom_attributes"),
+		EventTypes:       convertStringSet(d.Get("event_types").(*schema.Set)),
+		ObjectNamePrefix: d.Get("object_name_prefix").(string),
+		PayloadFormat:    d.Get("payload_format").(string),
+		Topic:            computedTopicName,
+	}
+
+	res, err := config.clientStorage.Notifications.Insert(bucket, storageNotification).Do()
+	if err != nil {
+		return fmt.Errorf("Error creating notification config for bucket %s: %v", bucket, err)
+	}
+
+	d.SetId(fmt.Sprintf("%s/notificationConfigs/%s", bucket, res.Id))
+
+	return resourceStorageNotificationRead(d, meta)
+}
+
+func resourceStorageNotificationRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	bucket := d.Get("bucket").(string)
```

Comment:
Since the bucket is already part of the id, you could avoid this d.Get here by using the return value from the parse function, which would then allow you to use an ImportStatePassthrough